### ISSUE TITLE
CrossEntropy: abort() on NaN softmax drops the diverged run's log

### ISSUE THREAD

**Issue by LeoBuron** [es-ude/OnDeviceTraining#446]
**Audit context:** found on 2026-09-17 while running the phase-2 optuna grid runner (es-ude/OnDeviceTraining-base-project#4) against `train_c_har_classifier` built from `develop`. Line numbers reference `es-ude/OnDeviceTraining@24d1760dbcd788d0d23bc24ee8d60ef93d8ae6fd`.

## Summary

`crossEntropyForwardFloat` calls `abort()` when a softmax output is non-finite. A run that diverges (a legitimate outcome of an LR/batch sweep) therefore dies with SIGABRT mid-epoch: the HAR harness's JSON log is left open without the epoch record and without `final`, so the divergence is not recorded anywhere, even though `examples/_shared/log_schema.py` was explicitly built to record it.

## Severity

| Axis | Value |
|---|---|
| Priority | Medium |
| Affected users | Subset: anyone sweeping learning rate or batch size, in particular the batch-size-scheduler experiments E0–E2 at `b_0 = 1` |
| Failure mode | Abort (loud), but the information is lost: partial log, no epoch record, exit code −6 indistinguishable from a crash |
| Blast radius | Experiment tooling / UX; training correctness is unaffected (a diverged run is invalid either way) |

## Root cause

https://github.com/es-ude/OnDeviceTraining/blob/24d1760dbcd788d0d23bc24ee8d60ef93d8ae6fd/src/loss_functions/CrossEntropy.c#L19-L26

```c
    for (size_t i = 0; i < n; i++) {
        float pi = p[i];

        if (!isfinite(pi)) {
            printf("NaN/Inf softmax at %zu\n", i);
            printf("%f\n", pi);
            abort();
        }
```

A library kernel terminates the process (and prints to stdout rather than through `Log.h`). Nothing above it gets a chance to react: `trainingRun` cannot return, the harness's `epochCallback` for the diverged epoch never fires, and `fclose(g_log_file)` (train_c.c:623) is only reached on normal completion, so the JSON object stays open.

This contradicts the schema's stated contract, https://github.com/es-ude/OnDeviceTraining/blob/24d1760dbcd788d0d23bc24ee8d60ef93d8ae6fd/examples/_shared/log_schema.py#L104-L107:

> A divergent training run (loss -> inf/nan) must stay a RECORDED finding, not a JSONDecodeError that drops the whole sweep.

For the C side that sanitizer is currently dead code: C never lives long enough to write a `nan`.

## Reproducer

`train_c_har_classifier` at the SHA above, Release build via the `examples` preset (`-ffp-contract=fast`), macOS, `Apple LLVM 21.0.0 (clang-2100.3.34.2)`, HAR data from `prepare_data.py`, run from the repo root:

```
EPOCHS=2 BATCH_SIZE=1 LR=0.01 MOMENTUM=0.9 SEED=0 SHUFFLE_SEED=0 \
RESHUFFLE=1 STEP_SIZE=1 MAX_BATCH_SIZE=661 LR_SCHEDULE=none BS_SCHEDULE=none \
LOG_PATH=/tmp/diverge.json ./build/examples/examples/har_classifier/train_c_har_classifier
```

stdout:

```
epoch 0: train_loss=0.7824 val_loss=1.6757 val_acc=0.6041 wall_s=5.90
NaN/Inf softmax at 0
nan
```

exit status 134 (SIGABRT, reported as −6 by a Python parent). `/tmp/diverge.json` ends after the epoch-0 record (`parameter_updates: 6617`); no epoch-1 record, no `final`, object not closed. Same with `BS_SCHEDULE=exp GAMMA=0.5` (dies in epoch 1 at batch 2). The cause is a batch-64 learning rate applied at batch 1; with `LR=0.0001` the same two configs finish with val_acc 0.692517 and 0.605442.

## Impact

- An external sweep driver sees a non-zero exit and a partial file. It cannot tell "this configuration diverged" (a result) from "the binary crashed" (a bug), so the trial is retried on every resume instead of being recorded once as diverged.
- The per-epoch record that would show *when* the loss blew up is never written; the `_sanitize_nonfinite` path in `log_schema.py` cannot do what its docstring promises for C logs.
- Any consumer of the library (not just the harness) can be killed from inside a loss kernel with no way to intercept it; the message goes to stdout, not through `Log.h`.

## Recommended fix

Two layers, both small:

**A. Library: never abort in a kernel.** Drop the `abort()`; let the non-finite value propagate (`fmaxf(NaN, 1e-7f)` yields `1e-7f` on most libms, so either clamp explicitly or accumulate `NaN` deliberately) and report through `Log.h` at most once per call. The loss becomes `NaN`/`Inf`, which is the honest result. Do the same wherever a twin check exists (grep for `isfinite` + `abort` under `src/`).

**B. Harness: record the divergence and close the log.** In `examples/har_classifier/train_c.c` (and the sibling harnesses), check `isfinite(info.trainLoss)` in `epochCallback` (or the eval loss) and, on divergence, still write the epoch record (with the `nan`), write `final` with `NaN` metrics or an explicit `"diverged": true`, `fclose`, print one line to stderr, and exit with a distinct non-zero code (e.g. 2) so drivers can classify it. `log_schema.py` already parses the bare `nan` tokens the C `printf("%.6f")` emits.

Option B alone (a finiteness check in `trainingRun` returning a status) would also work but touches the public API; A + B keeps the library pure and the policy in the harness.

## Verification

- Unit test: `crossEntropyForwardFloat` on a softmax tensor containing `NAN` returns a non-finite loss and the test process survives. Today the test process aborts.
- Harness test (`python/tests` or a `c-bit-parity`-style job): run the reproducer above; assert exit status is neither 134 nor 0, `load_log(path)` succeeds, the last `epochs[]` record has a non-finite `train_loss`, and `final` is present.

## Related

None found (searched issues for "NaN softmax", "abort", "divergence"). Origin: es-ude/OnDeviceTraining-base-project#4 (phase-2 runner), where the smoke grid now pins `LR=0.0001` for `BATCH_SIZE=1` to avoid the abort.
